File: data/EXP01/go/2467/2467.output_2.go
func mostProfitablePath(edges [][]int, bob int, amount []int) int {
    graph := make(map[int][]int)
    for _, edge := range edges {
        u, v := edge[0], edge[1]
        if _, ok := graph[u]; !ok {
            graph[u] = []int{}
        }
        if _, ok := graph[v]; !ok {
            graph[v] = []int{}
        }
        graph[u] = append(graph[u], v)
        graph[v] = append(graph[v], u)
    }
    
    var dfs func(node int, parent int) (int, int)
    dfs = func(node int, parent int) (int, int) {
        profit := 0
        gateCost := 0
        aliceProfit := math.MinInt32
        for _, child := range graph[node] {
            if child == parent {
                continue
            }
            childGateCost, childProfit := dfs(child, node)
            profit += childProfit
            gateCost += childGateCost
            aliceProfit = max(aliceProfit, childProfit)
        }
        if gateCost > amount[node] {
            diff := gateCost - amount[node]
            profit -= 2 * diff
            gateCost -= diff
            aliceProfit = max(aliceProfit-diff, 0)
        } else {
            diff := amount[node] - gateCost
            profit += diff
            aliceProfit += diff
        }
        if node == bob {
            gateCost /= 2
            profit /= 2
            aliceProfit /= 2
            aliceProfit += amount[node]
        }
        return gateCost, max(aliceProfit, 0)
    }
    
    gateCost, aliceProfit := dfs(0, -1)
    return aliceProfit + gateCost
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: data/EXP01/python/2467/2467.output_2.py
# @lc app=leetcode id=2467 lang=python3
from typing import List

class Solution:
    def mostProfitablePath(self, edges: List[List[int]], bob: int, amount: List[int]) -> int:
        graph = {}
        for edge in edges:
            u, v = edge
            if u not in graph:
                graph[u] = []
            if v not in graph:
                graph[v] = []
            graph[u].append(v)
            graph[v].append(u)
        
        def dfs(curr, prev, income):
            if not graph[curr]:  # current node is a leaf node
                return income
            
            max_income = -float('inf')
            for neighbor in graph[curr]:
                if neighbor == prev:  # skip the previous node
                    continue
                if income >= 0:  # Alice pays her share for opening the gate
                    new_income = income + amount[curr] // 2
                else:  # Alice receives her share for opening the gate
                    new_income = income + amount[curr]
                max_income = max(max_income, dfs(neighbor, curr, new_income))
            
            return max_income
        
        return dfs(0, bob, 0)